### PR TITLE
Prevent path traversal attacks on public content

### PR DESCRIPTION
**Pull request by jbdutton**
Mitigates path traversal attacks by using path.normalize to resolve any directory changes (`./` and `../`) and verifying that the path has one of the allowed prefixes in publicRoutes.

### CODE REVIEW

File: plugins/pencilblue/controllers/public.js [hunk 1]
```diff
@@ -44,6 +44,7 @@ module.exports = function PluginPublicContentControllerModule(pb) {
         var plugin          = this.pathVars.plugin;
         var postPluginPath  = this.pathVars.path;
         var pluginPublicDir = PluginService.getActivePluginPublicDir(plugin);
+        var publicRoutes = ['js/', 'css/', 'fonts/', 'img/', 'localization/', 'favicon.ico', 'docs/', 'dist/'];
```
Comment (lgenet) [inline plugins/pencilblue/controllers/public.js]: Should be const

File: plugins/pencilblue/controllers/public.js [hunk 2]
```diff
@@ -53,10 +54,20 @@ module.exports = function PluginPublicContentControllerModule(pb) {
...
 
         //serve up the content
-        var resourcePath = path.join(pluginPublicDir, postPluginPath);
-
-        //remove qsvars before loading files
-        this.reqHandler.servePublicContent(resourcePath.split('?')[0]);
+        //mitigates path traversal attacks by using path.normalize to resolve any
+        //directory changes (./ and ../) and verifying that the path has one of
+        //the prefixes in publicRoutes
+        var normalizedpath = path.normalize(postPluginPath).replace(/^(\.\.[\/\\])+/, '');
```
Comment (lgenet) [inline plugins/pencilblue/controllers/public.js]: I would say comment this rune but the text above should be good enough

File: plugins/pencilblue/controllers/public.js [hunk 3]
```diff
@@ -53,10 +54,20 @@ module.exports = function PluginPublicContentControllerModule(pb) {
...
+        
+        if (publicRoutes.some(prefix => normalizedpath.startsWith(prefix))) {
+            var fullpath = path.join(pluginPublicDir, normalizedpath);
+            
+            //remove qsvars before loading files
+            this.reqHandler.servePublicContent(fullpath.split('?')[0]);
+        } else {
+            this.reqHandler.serve404();
```
Comment (lgenet) [inline plugins/pencilblue/controllers/public.js]: Put return and this. On the same line. And remove else block

**Review by lgenet — APPROVED** [review]

File: plugins/pencilblue/controllers/public.js (hunk 1 above)
Comment (lgenet) [reply plugins/pencilblue/controllers/public.js]: Why expose docs

**Comment by jbdutton** [comment]
@lgenet I'm trying to match the current style of the code and not refactor everything at the moment.

**Pull request merged**